File: python/simple-linked-list/simple_linked_list.py
"""
(Singly) Linked List implementation with Node and EmpyListExeption classes
"""

class EmptyListException(Exception):
    """
    Exception for an empty linked list
    """
    pass


class Node:
    """
    Node object. The link in the linked list
    """
    def __init__(self, value):
        """
        Construct a new linked list node.
        :param value: The value for this node.
        """
        self.node_value = value
        self.node_next = None

    def value(self):
        """
        Return the value for this node.
        """
        return self.node_value

    def next(self):
        """
        Return a reference to the next item in the linked list.
        """
        return self.node_next


class LinkedList:
    """
    Linked list of Node objects
    """
    def __init__(self, values=None):
        """
        Create a new linked list with optional starting values
        :param values: list of values to add to the list
        """
        self.top = None
        self.index = None
        if values is not None:
            for value in values:
                self.push(value)

    def __iter__(self):
        """
        Setup the linked list for iteration
        :returns: object to iterate with
        """
        self.index = self.top
        return self
    
    def __next__(self):
        """
        Get the current item in the list and move to the next
        :returns: value at the current position in the list
        :raises: StopIteration if we get to the end of the list
        """
        if self.index is None:
            raise StopIteration
        
        current = self.index.value()
        self.index = self.index.next()
        return current

    def __len__(self):
        """
        Get the length of this linked list
        :returns: length of the linked list.
        """
        ret = 0
        current = self.top
        while current is not None:
            ret += 1
            current = current.next()
        return ret

    def head(self):
        """
        Return the node at the head of the list
        :returns: The node at the head of the list
        :raises: EmptyListException if list is empty.
        """
        if self.top is None:
            raise EmptyListException('The list is empty.')
        return self.top

    def push(self, value):
        """
        Push a new value to the head of the list.
        :param value: The value to add (as a node) to the head of the list.
        """
        new_node = Node(value)
        if self.top is not None:
            new_node.node_next = self.top
        
        self.top = new_node


    def pop(self):
        """
        Pop a value from the linked list at the head of the list.
        :returns: Value at the head of the list removing that node.
        :raises: EmptyListException if the list is empty.
        """
        if self.top is None:
            raise EmptyListException('The list is empty.')
        
        current_value = self.top.value()
        self.top = self.top.next()
        return current_value

    def reversed(self):
        """
        Reverse this linked list. Return the values as a list in reverse order
        :returns: A list of the linked list node values in reverse order.
        """
        current = self.top
        ret = []
        while current is not None:
            ret.insert(0, current.value())
            current = current.next()

        return ret

File: go/simple-linked-list/simple_linked_list.go
// Simple Linked list exercise.
package simplelinkedlist

import (
	"errors"
	"slices"
)

// Define the List and Element types here.
type List struct {
	head *Element
}

type Element struct {
	data int
	next *Element
}

// Create a brand new list and add the given elements to it.
// @param elements: zero or more elements to add to the new list.
// @returns: Reference to the new list.
func New(elements []int) *List {
	var ret List = List{head: nil}

	for _, elem := range elements {
		ret.Push(elem)
	}
	return &ret
}

// Return the number of items in the list (linear cost)
// @returns: The number of elements in the list.
func (l *List) Size() int {
	var count int = 0
	for elem := l.head; elem != nil; elem = elem.next {
		count++
	}
	return count
}

// Push a new item to the head of the list.
// @param element: The value to push
func (l *List) Push(element int) {
	node := Element{data: element, next: l.head}
	l.head = &node
}

// Return the value at the head of the list and remove it.
// @returns: Item at the head of the list.
// @raises: Error if the list is empty
func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("list is empty")
	}
	top := l.head
	l.head = top.next
	return top.data, nil
}

// Return the value at the head of the list.
// @returns: Item at head of list or error for empty list
// @raises: Error is empty list.
func (l *List) Peek() (int, error) {
	if l.head == nil {
		return 0, errors.New("list is empty")
	}

	return l.head.data, nil
}

// Returns the items in the list as a slice of int
// @returns: Slice of integer for the items in the list.
func (l *List) Array() []int {
	var ret []int = make([]int, l.Size())
	var index = 0
	for current := l.head; current != nil; current = current.next {
		ret[index] = current.data
		index++
	}
	slices.Reverse(ret)
	return ret
}

// Return a reversed copy of this list.
// @returns: New list of same items but in reverse order
func (l *List) Reverse() *List {
	var nums []int = l.Array()
	slices.Reverse(nums)
	return New(nums)
}
